game/quake2_md2.ksy

```diff
@@ -0,0 +1,249 @@
+    type: u4
+  - id: ofs_frames
+    type: u4
+  - id: ofs_gl_cmds
+    type: u4
+  - id: ofs_eof
+    type: u4
+instances:
+  skins:
+    pos: ofs_skins
+    size: 64
+    type: strz
+    encoding: ascii
+    repeat: expr
+    repeat-expr: num_skins
+  tex_coords:
+    pos: ofs_tex_coords
+    type: tex_point
+    repeat: expr
+    repeat-expr: num_tex_coords
+  triangles:
+    pos: ofs_triangles
+    type: triangle
+    repeat: expr
+    repeat-expr: num_triangles
+  frames:
+    pos: ofs_frames
+    type: frame
+    repeat: expr
+    repeat-expr: num_frames
+  gl_cmds:
+    pos: ofs_gl_cmds
+    type: gl_cmd
+    repeat: until
+    repeat-until: _.cmd_num_vertices == 0
+types:
+  tex_point:
+    seq:
+      - id: u_px
+        type: u2
+      - id: v_px
+        type: u2
+    instances:
+      u_normalized:
+        value: u_px.as<f4> / _root.skin_width_px
+      v_normalized:
+        value: v_px.as<f4> / _root.skin_height_px
+  triangle:
+    seq:
+      - id: vertex_indices
+        type: u2
+        repeat: expr
+        repeat-expr: 3
+      - id: tex_point_indices
+        type: u2
+        repeat: expr
+        repeat-expr: 3
+  vec3f:
+    seq:
+      - id: x
+        type: f4
+      - id: y
+        type: f4
+      - id: z
+        type: f4
+  compressed_vec:
+    seq:
+      - id: x_compressed
+        type: u1
+      - id: y_compressed
+        type: u1
+      - id: z_compressed
+        type: u1
+    instances:
+      x:
+        value: x_compressed * _parent._parent.scale.x + _parent._parent.translate.x
+      y:
+        value: y_compressed * _parent._parent.scale.y + _parent._parent.translate.y
+      z:
+        value: z_compressed * _parent._parent.scale.z + _parent._parent.translate.z
+  vertex:
+    seq:
+      - id: position
+        type: compressed_vec
+      - id: normal_index
+        type: u1
+        doc: `normal = bytedirs[normal_index]`
+        doc-ref: |
+          https://github.com/skullernet/q2pro/blob/master/src/common/math.c#L80
+          from Quake anorms.h
+  frame:
+    seq:
+      - id: scale
+        type: vec3f
+      - id: translate
+        type: vec3f
+      - id: name
+        size: 16
+        type: strz
+        encoding: ascii
+      - id: vertices
+        type: vertex
+        repeat: expr
+        repeat-expr: _root.vertices_per_frame
+  gl_cmd:
+    seq:
+      - id: cmd_num_vertices
+        type: s4
+      - id: vertices
+        type: gl_vertex
+        repeat: expr
+        repeat-expr: num_vertices
+    instances:
+      num_vertices:
+        value: 'cmd_num_vertices < 0 ? -cmd_num_vertices : cmd_num_vertices'
+      primitive:
+        value: 'cmd_num_vertices > 0 ? gl_primitive::triangle_strip : gl_primitive::triangle_fan'
+  gl_vertex:
+    seq:
+      - id: tex_coords_normalized
+        type: f4
+        repeat: expr
+        repeat-expr: 2
+      - id: vertex_index
+        type: u4
+enums:
+  gl_primitive:
+    0: triangle_strip
+    1: triangle_fan
+  anim_start_index:
```

Comment:
I have added this machine-readable version of part of the animation table. I would also like to include machine-readable data for how many frames each anim has, but I can't see how to do this in Kaitai. For example, the three pain anims all have 4 frames, which would clash if I tried to use another enum for this purpose. Is there any way to include constant key-value definitions where several keys might share the same value?

Reply:
@jdjakub:

> For example, the three pain anims all have 4 frames, which would clash if I tried to use another enum for this purpose.

Yes, an enum in Kaitai Struct is a strictly bidirectional mapping ([one-to-one correspondence](https://en.wikipedia.org/wiki/Bijection)).

> Is there any way to include constant key-value definitions where several keys might share the same value?

You can use [value instances](https://doc.kaitai.io/user_guide.html#_value_instances) to define constants:

```ksy
instances:
  my_const:
    value: 4
```

Technically you can also declare arrays there (though I'm not sure whether it makes sense here):

```ksy
instances:
  my_array:
    value: |
      [2, 3, 5, 7, 11, 13]
```

Another option (but it depends on what you mean by machine-readable) may be to add a custom YAML key somewhere. KS compiler ignores any properties starting with a dash `-` (see https://doc.kaitai.io/user_guide.html#orig-id), so this can be used to add arbitrary metadata. But these keys won't be provided in generated parsers.

Reply:
Thanks. I have replaced the enum with three value instances and updated the doc to refer to them. If it's OK then please go ahead and merge, I want to be able to look it up on the website :)